<a href="https://colab.research.google.com/github/fasthill/ML-DL-study-alone/blob/main/7-2%20%EC%8B%AC%EC%B8%B5%20%EC%8B%A0%EA%B2%BD%EB%A7%9D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 심층 신경망

## 2개의 층

In [1]:
import numpy as np

from sklearn.model_selection import train_test_split
from tensorflow import keras
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense, Flatten

In [2]:
(train_input, train_target), (test_input, test_target) = fashion_mnist.load_data()

4431872/4422102 [==============================] - 0s 0us/step


In [3]:
train_scaled = train_input / 255
test_scaled = test_input / 255
train_scaled = train_scaled.reshape(-1, (28*28))
test_scaled = test_scaled.reshape(-1, (28*28))

In [4]:
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42)

## 심층 신경망 만들기

In [5]:
dense1 = Dense(100, activation='sigmoid', input_shape=(784,))
dense2 = Dense(10, activation='softmax')

In [6]:
model = Sequential([dense1, dense2])

In [7]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               78500     
                                                                 
 dense_1 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


#### OutputShape = (None, 100) : None: batchsize 표시. none으로 되어 있어서, 변경이 자유롭게 됨을 표시. 100: 100개의 노드 출력. <br>
78500 : 784개 입력 \* 은닉층 노드수 100에 해당하는 가중치 w의 갯수 + 100개의 bias(절편) ( 784 × 100 + 100 )
#### OutputShape = (None, 10) : 10개의 노드 출력. <br>
1010 : 입력으로 100개 은닉층 \* 출력층 노드수 10에 해당하는 가중치 w의 갯수 + 10개의 bias(절편) ( 100 × 10 + 10 )
#### Total params : 79510 = 78500 + 1010

## 층을 추가하는 다른 방법

In [8]:
model = keras.Sequential([
                          keras.layers.Dense(100, activation='sigmoid', input_shape=(784,), name='hidden'),
                          Dense(10, activation='softmax', name='output')], 
                         name='패션 MNIST 모델')

In [9]:
model.summary()

Model: "패션 MNIST 모델"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 hidden (Dense)              (None, 100)               78500     
                                                                 
 output (Dense)              (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


In [10]:
model = keras.Sequential(name='Sequen_1')
model.add(keras.layers.Dense(100, activation='sigmoid', input_shape=(784,), name='hidden_1'))
model.add(keras.layers.Dense(10, activation='softmax', name='output'))

가장 많이 사용하는 모델 구성 방식. if문 등을 사용하여 프로그램 식으로 구성할 수 있음.

In [11]:
model.summary()

Model: "Sequen_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 hidden_1 (Dense)            (None, 100)               78500     
                                                                 
 output (Dense)              (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


In [12]:
model.compile(loss='sparse_categorical_crossentropy', metrics='accuracy')

model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 8s 3ms/step - loss: 0.5677 - accuracy: 0.8068
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.4073 - accuracy: 0.8540
Epoch 3/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3741 - accuracy: 0.8638
Epoch 4/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3522 - accuracy: 0.8720
Epoch 5/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3354 - accuracy: 0.8774


## 렐루 활성화 함수

In [13]:
model = Sequential()
model.add(Flatten(input_shape=(28,28)))
model.add(Dense(100, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [14]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense_2 (Dense)             (None, 100)               78500     
                                                                 
 dense_3 (Dense)             (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


In [15]:
model.compile(loss='sparse_categorical_crossentropy', metrics=['accuracy','mse'])

In [16]:
(train_input, train_target), (test_input, test_target) = fashion_mnist.load_data()
train_scaled = train_input / 255.0
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42)  # Flatten method를 사용하기 위하여 원래의 shape로 구성.

In [17]:
train_scaled.shape

(48000, 28, 28)

In [18]:
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.5343 - accuracy: 0.8109 - mse: 27.6957
Epoch 2/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3951 - accuracy: 0.8583 - mse: 27.7035
Epoch 3/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3546 - accuracy: 0.8726 - mse: 27.7057
Epoch 4/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3344 - accuracy: 0.8818 - mse: 27.7072
Epoch 5/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3193 - accuracy: 0.8852 - mse: 27.7081


In [19]:
model.evaluate(val_scaled, val_target) # loss, accuracy .compile에서 정의한 metrics 내용 출력

375/375 [==============================] - 1s 3ms/step - loss: 0.3594 - accuracy: 0.8791 - mse: 27.5963


[0.359444797039032, 0.8790833353996277, 27.596250534057617]

## 옵티마이저

In [20]:
model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics='accuracy')

In [21]:
model.fit(train_scaled, train_target, batch_size=200, epochs=5) # default batch_size=32

Epoch 1/5
240/240 [==============================] - 1s 3ms/step - loss: 0.2642 - accuracy: 0.9057
Epoch 2/5
240/240 [==============================] - 1s 3ms/step - loss: 0.2575 - accuracy: 0.9081
Epoch 3/5
240/240 [==============================] - 1s 3ms/step - loss: 0.2553 - accuracy: 0.9085
Epoch 4/5
240/240 [==============================] - 1s 3ms/step - loss: 0.2538 - accuracy: 0.9093
Epoch 5/5
240/240 [==============================] - 1s 3ms/step - loss: 0.2527 - accuracy: 0.9092


In [22]:
sgd = keras.optimizers.SGD() # default 값만 사용하면 optimizers='sgd'와 동일
model.compile(optimizer=sgd, loss='sparse_categorical_crossentropy', metrics='accuracy')

In [23]:
sgd = keras.optimizers.SGD(learning_rate=0.1) # 객체로 만들어서 parameter를 이용하여 객체로 만들어서 사용할 수 있음.
model.compile(optimizer=sgd, loss='sparse_categorical_crossentropy', metrics='accuracy')

In [24]:
sgd = keras.optimizers.SGD(momentum=0.9, nesterov=True) # 객체로 만들어서 parameter를 이용하여 객체로 만들어서 사용할 수 있음.
model.compile(optimizer=sgd, loss='sparse_categorical_crossentropy', metrics='accuracy')

In [25]:
adagrad = keras.optimizers.Adagrad()
model.compile(optimizer=adagrad, loss='sparse_categorical_crossentropy', metrics='accuracy')

In [26]:
rmsprop = keras.optimizers.RMSprop()
model.compile(optimizer=rmsprop, loss='sparse_categorical_crossentropy', metrics='accuracy')

In [27]:
model = keras.Sequential()
model.add(keras.layers.Flatten(input_shape=(28, 28)))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(10, activation='softmax'))

In [28]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics='accuracy')
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 [==============================] - 5s 3ms/step - loss: 0.5195 - accuracy: 0.8179
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3953 - accuracy: 0.8593
Epoch 3/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3508 - accuracy: 0.8730
Epoch 4/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3255 - accuracy: 0.8810
Epoch 5/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3061 - accuracy: 0.8876


In [29]:
model.evaluate(val_scaled, val_target)

375/375 [==============================] - 1s 2ms/step - loss: 0.3536 - accuracy: 0.8727


[0.3536130487918854, 0.8726666569709778]